### PR TITLE
use fast path in blendSegment to bump up FPS

### PR DESCRIPTION
**Pull request by DedeHai**
currently blendSegment is "the bottleneck" in the rendering pipeline. I ran a breakdown of times (in µs):

- total=8958 
- effect rendering: 6156 
- show: 2677 = **_blend 1359_** + pixelstobus 1318

this is for a single segment. blending is slow for a "just copy the segment buffer"

with this PR it becomes:

- total=8091 
- effect rendering: 6536 (fluctuates)
- show: 1425 = **_blend  105_** + pixelstobus 1320 

Improvements in FPS:
- 2 overlapping segments: 76.6FPS -> 84.5FPS
- single segment: 103FPS -> 110FPS

i.e. up to 10% faster.

test run on C3, 2 outputs 256 pixels each, 32x16 matrix using PS-Fire

## Summary by CodeRabbit

* **Refactor**
  * Faster, smoother LED animations — optimized fast-path for fade-style transitions when grouping and mirroring are not used.
  * Reduced per-frame computation by skipping unnecessary progress math and eliminating duplicated work.
  * More efficient pixel buffer handling and quicker frame clears for snappier updates on strips and matrices.

### CODE REVIEW

**Comment by DedeHai** [comment]
just FYI: ported this improvement over to my WLEDbus driver branch, I do get ~100FPS for the 2 overlapping segments which come in at 84.5FPS using NPB.

**Comment by softhack007** [comment]
@DedeHai i'll use your PR as a test-case for the new rules for the rabbit 🔢 

@coderabbitai please re-review this PR, using the new coding and AI rules from the `main` branch (applicable via .coderabbit.yaml). Summarize good practices, and create a prioritized list of possible improvements (if any).

File: wled00/FX_fcn.cpp [hunk 1]
```diff
@@ -1643,7 +1698,7 @@ void WS2812FX::show() {
 
   if (realtimeMode == REALTIME_MODE_INACTIVE || useMainSegmentOnly || realtimeOverride > REALTIME_OVERRIDE_NONE) {
     // clear frame buffer
-    for (size_t i = 0; i < totalLen; i++) _pixels[i] = BLACK; // memset(_pixels, 0, sizeof(uint32_t) * getLengthTotal());
+    memset(_pixels, 0, sizeof(uint32_t) * totalLen);
```
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX_fcn.cpp:1703]: this is a good point actually. While it works it is fragile and not guaranteed to do so in the future. please check the codebase for any other instances of memset or other functions that may use 8bit access on 32bit allocated pixelbuffers.

[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX_fcn.cpp:1703]: given the fact that the code using memcopy and memset works flawlessly on an ESP32 with its buffers in IRAM and a violation will result in an immediate crash in testing - as buffers will be allocated in this ram region: how likely is it that this may break in a future IDF version?

[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX_fcn.cpp:1703]: all pixel buffers are currently in 32bit accessible RAM, this is by design and intentional. your assessment is wrong that it will not fit - it almost certainly fits up to 1000 LEDs. meaning in tests, memset and memcopy do work and if assembly code is used it is highly unlikely to change in the future so the current approach seems to be fine, even though it is a risk.

File: wled00/FX_fcn.cpp [hunk 2]
```diff
@@ -1398,14 +1399,69 @@ void WS2812FX::blendSegment(const Segment &topSegment) const {
...
+            *p = color_blend(*p, segblend(c_a, *p), opacity);
+          }
+        }
+      } else { // transposed
+        for (int y = 0; y < height; y++) {
+          for (int x = 0; x < width; x++) {
+            int px = topSegment.reverse   ? (width - x - 1)  : x;
+            int py = topSegment.reverse_y ? (height - y - 1) : y;
+            size_t idx = XY(topSegment.start + py, topSegment.startY + px); // Swapped X/Y
+            _pixels[idx] = color_blend(_pixels[idx], segblend(topSegment.getPixelColorRaw(x + y * width), _pixels[idx]), opacity);
```
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX_fcn.cpp]: the slow path traverses the segment, reads the pixel, then transposes/inverts coordinates, then writes. the fast path does the same does it not? I don't see the error.

**Comment by DedeHai** [comment]
@coderabbitai please put together a test plan: what settings to use and what to look for to catch any edge cases

**Comment by DedeHai** [comment]
I ran most of the tests the rabbit suggested. All looking good.
Ready to merge.

**Review by willmmiles — APPROVED** [review]
I've been meaning to get to this for a while!  This is a step in the right direction.  I think there are some optimized blend primitive functions that could be factored out and composed to make fast paths for even more cases.

(The general principle is called "hoisting" -- moving conditionals outside of the hot loops, trading off some code replication for speed.  I think it's a good call for our main render loops.)
